src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs

```diff
@@ -378,7 +378,11 @@ private async Task HeartbeatAsync(CancellationToken cancellationToken)
 
                 newDescription = newDescription.With(reasonChanged: "Heartbeat", lastHeartbeatTimestamp: DateTime.UtcNow);
 
-                SetDescription(newDescription);
+                lock (_lock)
```

Comment:
Why was this change required?

Reply:
In case if we cancel the current isMaster check after these lines: https://github.com/DmitryLukyanov/mongo-csharp-driver/blob/9e90b5b90479cc69f1bb78d550e0ac5e62c6733e/src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs#L329-L332 but before publishing it. In this case, without this line we will still publish the cancelled isMaster data

Reply:
Maybe a better option, move lines 329-332 to line 383 for consistency

Reply:
NOTE: in the current implementation, the thrown exception will be caught here https://github.com/DmitryLukyanov/mongo-csharp-driver/blob/9e90b5b90479cc69f1bb78d550e0ac5e62c6733e/src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs#L211-L214

Reply:
I will defer to you and @rstam

Reply:
I'm fine with adding this lock statement.